### DISCUSSION TITLE
Unsuccessfully tried to record videos with USB Basler camera

### DISCUSSION THREAD

**Discussion by KCl94** [bonsai-rx/bonsai#1193] [Q&A]
Hi everyone,

First time posting here. I have been trying for some time now to set up a simple recording system with one grayscale Basler camera. From what I have found on previous posts on this site, I was able to create a .pfs file through Pylon 5.1 for the PylonCapture node. I then added a MemberSelector node (called Image), which I got from other users working with the Basler cameras as well (I couldn’t figure out how to extract this property myself). Lastly, everything is connected to the ffmpeg-dependent VideoWriter node. The ffmpeg and Pylon libraries are installed, too.

The main issue I have been having is that I just can’t get a video to save. Previously I was trying to use the other VideoWriter node but 1) .avi files are massive and 2) all the videos drop A LOT of frames (all of the videos I would obtain were minutes short). I have found on other posts that I could use the ImageWriter node to save a video, but I still can’t succeed at this seemingly simple task. Could you guys please guide me in the right direction? I am hoping to eventually set up four cameras on separate boxes and then, if possible, run them all in the same workflow but independently from each other. I would much prefer to save my videos in .mp4 format. Additionally, I just discovered today that, after updating to Bonsai 2.7.2, an old error “External component has thrown an exception” is back when trying to start the workflow. I had this error before, but I solved it by downgrading to Pylon 5.1 and creating a Bonsai-compatible parameter file, as suggested [here](https://groups.google.com/g/bonsai-users/c/vpf62pjF45Q/m/H5Xwr7N_CwAJ). Any idea as to why this error is back? I am attaching a screenshot that I hope will be useful in getting to determine the root of the problem. Thanks a lot, in advance!

![BonsaiScreenshots](https://user-images.githubusercontent.com/109825179/207098320-9e885d08-5e96-4e16-b885-d1fd6b239ea6.png)

**Comment by bruno-f-cruz** [comment]
What version of the pylon package did you install?

**Reply by KCl94** [reply]
Hi! I downgraded to 5.1. This worked for me until the latest Bonsai update. Unfortunately, by "worked for me" I really mean that the "External component has thrown an exception" error wouldn't pop up. Even then I couldn't save a video, and the command prompt (console) is almost always blank.

**Reply by bruno-f-cruz** [reply]
I mean the Bonsai Pylon package. Not the Pylon Viewer.

**Reply by KCl94** [reply]
Oh, sorry. That is 2.7.2

**Reply by bruno-f-cruz** [reply]
That's the Bonsai version, what I really need is the package version. You can check it  by going into the package manager. In my case, I am running the 0.3.0.
![image](https://user-images.githubusercontent.com/7049351/207112191-f3e17206-81a2-4492-8485-6f6c1f0f06e8.png)

**Reply by KCl94** [reply]
Haha! My apologies. I have the same version you just showed me, yes

**Reply by bruno-f-cruz** [reply]
As the package description suggests, try to install the pylon runtime version 7.1.0. That should fix the bug. 
Regarding your other problem, can you save the videos using a non-FFMPEG VideoWriter node?

**Reply by KCl94** [reply]
Well, previously I was trying to use the non-ffmpeg VideoWriter node but the .avi files tend to be massive and, more importantly, all of the videos I would obtain were minutes shorter than expected. The recording sessions are supposed to be one hour long, but I would be as far behind as 50–55-minute videos.

**Reply by bruno-f-cruz** [reply]
Was the number of frames correct? Consider reading the discussion in #1112. 
Regarding the ffmpeg, I am not sure which codec combination you are trying to use, can you share the workflow?
Thanks!

**Comment by KCl94** [comment]
Here is the link to the simple workflow. I apologize if I am ignoring something super clear, I am still learning this program: 
https://drive.google.com/file/d/1OIfKxmEMu-Vqggz6ntCk2xtXAMQ4FPAf/view?usp=sharing

**Reply by bruno-f-cruz** [reply]
Would you mind attaching it as a `.zip` file to this post? Otherwise, I feel you might receive the occasional email asking access to the link you just shared :)

**Reply by KCl94** [reply]
[TobeShared.zip](https://github.com/bonsai-rx/bonsai/files/10210797/TobeShared.zip)
Good idea!

**Reply by bruno-f-cruz** [reply]
That's bizarre, the workflow you sent works perfectly on my computer. Try to update your Pylon runtime and try again. Can you also tell me what is the resolution of your input image?

**Reply by KCl94** [reply]
Downloading 7.1.0 as we speak. So, were you able to save a video? No issues at all?

**Reply by bruno-f-cruz** [reply]
Yes.

**Reply by KCl94** [reply]
So, I was able to run the program after the Pylon update, but the main issue still persists. I got an unplayable video. It is approximately 43 mb in size, so it shouldn't be empty! But I can't open it with neither video player on my computer. This is what I see: 

![Brokenoutput](https://user-images.githubusercontent.com/109825179/207125234-b03ca02b-719d-44df-827e-82c71df1428b.png)

Do you think I am missing something under "output arguments" for the ffmpeg VideoWriter node?

The expected resolution is 4024 x 3036, as these are the default on the Basler camera I have.

**Reply by bruno-f-cruz** [reply]
I might be wrong, but I think the default ffmpeg codec for .mp4 only supports 2048x2048... can you try to resize your image (use the `Resize` operator) and give it a try?

**Reply by KCl94** [reply]
Hi again! So, I just resized the image to 2048x2048, as you suggested, and this works! Unfortunately, this only works with one camera. When I multiply this workflow for four cameras, the videos are of irregular lengths (a two-minute recording gave me four videos of different time lengths).
Curious, do I need to create a separate .pfs file for each camera?

**Comment by bruno-f-cruz** [comment]
How are you triggering the cameras? With a common trigger signal?

**Reply by KCl94** [reply]
Yeah, I simply press the Start button for now and recordings start immediately. My hope is to eventually create a workflow such that I can start the program and launch "preview" windows that would be waiting on my command (a key press) to start and stop recording after another key press

**Comment by bruno-f-cruz** [comment]
Just to clarify, you are sending a synchronized pulse to the gpio of every
camera? How long after starting bonsai do you turn this pulse on?

**Reply by KCl94** [reply]
Oh no, I am not that far yet. All I do is press the Start button on Bonsai. No pulses yet.

**Comment by bruno-f-cruz** [comment]
If you are not forcing any synchronization between cameras they will likely
output a different number of frames and hence different video duration.

**Reply by KCl94** [reply]
That makes a lot of sense, actually. Could you guide me in the right direction?
Also, do you know if fixing this synchronization issue would also alleviate the memory strain? My Intel i7-10700k CPU @ 3.9GHz 64 GB RAM computer was saturated during the latest attempt. 

Thanks a lot for your help so far, btw :)
